File: C#.Net/JDI/Web/JDI Web/Selenium/Elements/Complex/Table/Rows.cs
using System.Collections.Generic;
using System.Linq;
using JDI_Web.Selenium.Elements.Complex.Table.Interfaces;
using OpenQA.Selenium;
using JDI_Commons;
using static Epam.JDI.Core.Settings.JDISettings;
using static Epam.JDI.Core.ExceptionUtils;
using static JDI_Web.Utils.WebExtensions;

namespace JDI_Web.Selenium.Elements.Complex.Table
{
    public class Rows : TableLine
    {
        public Rows()
        {
            HasHeader = false;
            ElementIndex = ElementIndexType.Nums;
            HeadersLocator = By.XPath(".//tr/td[1]");
            DefaultTemplate = By.XPath(".//tr[{0}]/td");
        }

        protected override IList<IWebElement> GetFirstLine => Table.Columns.GetLineAction(1);

        public override Dictionary<string, Dictionary<string, ICell>> Get()
        {
            return Headers.ToDictionary(key => key, GetRow);
        }

        ///

        public IList<string> GetRowValue(string rowName)
        {
            return ActionWithException(() => GetLineAction(rowName).Select(el => el.Text).ToList(),
                ex => $"Can't Get Row '{rowName}'. Reason: {ex}");
        }

        public Dictionary<string, string> GetRowAsText(string rowName)
        {
            return GetRow(rowName).ToDictionary(pair => pair.Key, pair => pair.Value.GetText);
        }

        public Dictionary<string, ICell> CellsToRow(ICollection<ICell> cells)
        {
            return cells.ToDictionary(
                    cell => Headers[cell.ColumnNum - 1],
                    cell => cell);
        }

        public Dictionary<string, ICell> GetRow(int rowNum)
        {
            if (Count < 0 || Count < rowNum || rowNum <= 0)
                throw Exception($"Can't Get Row '{rowNum}'. [num] > ColumnsCount({Count}).");
            return ActionWithException(() =>
            {
                var colsCount = Table.Columns.Count;
                var webRow = Timer.GetResultByCondition(() => GetLineAction(rowNum), els => els.Count >= colsCount);
                if (webRow == null)
                    throw Exception($"Table has only {GetLineAction(rowNum).Count} columns " +
                                    $"but expected at least {colsCount}");
                var result = new Dictionary<string, ICell>();
                if (webRow.Count == colsCount)
                {
                    AddCols(result, Table.Columns.Headers, webRow, rowNum);
                    return result;
                }
                AddCols(result, Table.Columns.AllHeaders, webRow, rowNum);
                var simplifiedHeaders = Table.Columns.Headers.Select(Simplify);
                return result.Where(el => simplifiedHeaders.Contains(Simplify(el.Key))).ToDictionary();
            }, ex => $"Can't Get Row '{rowNum}'. Reason: {ex}");
        }
        private void AddCols(Dictionary<string, ICell> result, IList<string> headers, IList<IWebElement> webRow, int rowNum)
        {
            for (var i = 0; i < headers.Count; i++)
            {
                var label = headers[i];
                if (!result.ContainsKey(label))
                    result.Add(label, Table.Cell(webRow[i], new Column(i + 1), new Row(rowNum)));
            }
        }

        public IList<string> GetRowValue(int rowNum)
        {
            if (Count < 0 || Count < rowNum || rowNum <= 0)
                throw Exception($"Can't Get Row '{rowNum}'. [num] > ColumnsCount({Count}).");
            return ActionWithException(() => GetLineAction(rowNum).Select(el => el.Text).ToList(),
                ex => $"Can't Get Row '{rowNum}'. Reason: {ex}");
        }

        public Dictionary<string, string> GetRowAsText(int rowNum)
        {
            return GetRow(rowNum).ToDictionary(pair => pair.Key, pair => pair.Value.GetText);
        }

        public Dictionary<string, ICell> GetRow(string rowName)
        {
            return ActionWithException(() =>
            {
                var colsCount = Table.Columns.Count;
                var webRow = Timer.GetResultByCondition(() => GetLineAction(rowName), els => els.Count >= colsCount);
                if (webRow == null)
                    throw Exception($"Table has only {GetLineAction(rowName).Count} columns " +
                                    $"but expected at least {colsCount}");
                var result = new Dictionary<string, ICell>();
                if (webRow.Count == colsCount)
                {
                    AddCols(result, Table.Columns.Headers, webRow, rowName);
                    return result;
                }
                AddCols(result, Table.Columns.AllHeaders, webRow, rowName);
                return result.Where(el => Table.Columns.Headers.Contains(el.Key)).ToDictionary();
            }, ex => $"Can't Get Row '{rowName}'. Reason: {ex}");
        }
        private void AddCols(Dictionary<string, ICell> result, IList<string> headers, IList<IWebElement> webRow, string rowName)
        {
            for (var i = 0; i < headers.Count; i++)
                result.Add(headers[i], Table.Cell(webRow[i], new Column(name: headers[i]), new Row(name: rowName)));
        }
    }
}

File: Java/JDI/jdi-uitest-win/src/main/java/com/epam/jdi/uitests/win/winnium/elements/complex/table/Rows.java
package com.epam.jdi.uitests.win.winnium.elements.complex.table;

import com.epam.commons.linqinterfaces.JFuncTTREx;
import com.epam.commons.map.MapArray;
import com.epam.jdi.uitests.core.interfaces.common.IText;
import com.epam.jdi.uitests.core.interfaces.complex.interfaces.Column;
import com.epam.jdi.uitests.core.interfaces.complex.interfaces.ICell;
import com.epam.jdi.uitests.core.interfaces.complex.interfaces.IRow;
import com.epam.jdi.uitests.core.interfaces.complex.interfaces.Row;
import org.openqa.selenium.WebElement;

import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

import static com.epam.jdi.uitests.core.settings.JDISettings.exception;

public class Rows extends TableLine implements IRow {
    public Rows(Table table) {
        super(table);
    }

    @Override
    protected List<WebElement> getFirstLine() {
        return ((Columns)table.columns()).getLineAction(1);
    }

    @Override
    public MapArray<String, ICell> getRow(String rowName) {
        try {
            int colsCount = table.columns().count();
            List<String> headers = table.columns().headers();
            return new MapArray<>(colsCount, headers::get,
                    value -> table.cell(new Column(headers.get(value)), new Row(rowName)));
        } catch (Exception | Error ex) {
            throw throwRowsException(rowName, ex.getMessage());
        }
    }

    @Override
    public List<String> getRowValue(String rowName) {
        return getLineAction(rowName).stream().map(WebElement::getText).collect(Collectors.toList());
    }

    @Override
    public MapArray<String, String> getRowAsText(String rowName) {
        return getRow(rowName).toMapArray(IText::getText);
    }

    private MapArray<String, MapArray<String, ICell>> withValueByRule(Column column,
                                                                      JFuncTTREx<String, String, Boolean> func) {
        Collection<String> rowNames = column.hasName()
                ? table.columns().getColumnAsText(column.getName()).where(func).keys()
                : table.columns().getColumnAsText(column.getNum()).where(func).keys();
        return new MapArray<>(rowNames, key -> key, this::getRow);
    }

    @Override
    public final MapArray<String, MapArray<String, ICell>> withValue(String value, Column column) {
        return withValueByRule(column, (key, val) -> val.equals(value));
    }

    @Override
    public final MapArray<String, MapArray<String, ICell>> containsValue(String value, Column column) {
        return withValueByRule(column, (key, val) -> val.contains(value));
    }

    @Override
    public final MapArray<String, MapArray<String, ICell>> matchesRegEx(String regEx, Column column) {
        return withValueByRule(column, (key, val) -> val.matches(regEx));
    }

    @Override
    public MapArray<String, ICell> getRow(int rowNum) {
        if (count() < 0 || count() < rowNum || rowNum <= 0)
            throw exception("Can't Get Row '%s'. [num] > ColumnsCount(%s).", rowNum, count());

        int colsCount = table.columns().count();

        return new MapArray<>(colsCount,
                key -> table.columns().headers().get(key),
                value -> table.cell(new Column(value + 1), new Row(rowNum)));
    }

    @Override
    public List<String> getRowValue(int rowNum) {
        return getLineAction(rowNum).stream().map(WebElement::getText).collect(Collectors.toList());
    }

    @Override
    public MapArray<String, String> getRowAsText(int rowNum) {
        return getRow(rowNum).toMapArray(IText::getText);
    }

    @Override
    public MapArray<String, MapArray<String, ICell>> get() {
        return new MapArray<>(headers(), key -> key, this::getRow);
    }

    private RuntimeException throwRowsException(String rowName, String ex) {
        return exception("Can't Get Row '%s'. Reason: %s", rowName, ex);
    }
}
